src/sign.rs

```diff
@@ -547,78 +516,255 @@ impl CoseSign1 {
-        let map: HeaderMap = SignatureAlgorithm::ES256.into();
-        assert_eq!(
-            map_to_empty_or_serialized(&map).unwrap(),
-            [0xa1, 0x01, 0x26]
-        );
-
-        let map: HeaderMap = SignatureAlgorithm::ES384.into();
-        assert_eq!(
-            map_to_empty_or_serialized(&map).unwrap(),
-            [0xa1, 0x01, 0x38, 0x22]
-        );
-
-        let map: HeaderMap = SignatureAlgorithm::ES512.into();
-        assert_eq!(
-            map_to_empty_or_serialized(&map).unwrap(),
-            [0xa1, 0x01, 0x38, 0x23]
-        );
-    }
+    #[cfg(feature = "key_openssl_pkey")]
+    mod generic {
+        use crate::sign::*;
+        use openssl::pkey::{PKey, Private, Public};
+
+        // Public domain work: Pride and Prejudice by Jane Austen, taken from https://www.gutenberg.org/files/1342/1342.txt
+        const TEXT: &[u8] = b"It is a truth universally acknowledged, that a single man in possession of a good fortune, must be in want of a wife.";
+
+        #[test]
+        fn map_serialization() {
+            // Empty map
+            let map: HeaderMap = HeaderMap::new();
+            assert_eq!(map_to_empty_or_serialized(&map).unwrap(), []);
+
+            // Checks that the body_protected field will be serialized correctly
+            let map: HeaderMap = SignatureAlgorithm::ES256.into();
+            assert_eq!(
+                map_to_empty_or_serialized(&map).unwrap(),
+                [0xa1, 0x01, 0x26]
+            );
+
+            let map: HeaderMap = SignatureAlgorithm::ES384.into();
+            assert_eq!(
+                map_to_empty_or_serialized(&map).unwrap(),
+                [0xa1, 0x01, 0x38, 0x22]
+            );
+
+            let map: HeaderMap = SignatureAlgorithm::ES512.into();
+            assert_eq!(
+                map_to_empty_or_serialized(&map).unwrap(),
+                [0xa1, 0x01, 0x38, 0x23]
+            );
+        }
 
-    #[test]
-    fn map_with_duplicates() {
-        // Check that HeaderMaps with duplicate entries emit error
-        // {1: 42, 2: 42}
-        let test = [0xa2, 0x01, 0x18, 0x2A, 0x02, 0x18, 0x2A];
-        let map: HeaderMap = serde_cbor::from_slice(&test).unwrap();
-        assert_eq!(
-            map.get(&CborValue::Integer(1)),
-            Some(&CborValue::Integer(42))
-        );
-        assert_eq!(
-            map.get(&CborValue::Integer(2)),
-            Some(&CborValue::Integer(42))
-        );
-
-        // {1: 42, 2: 42, 1: 43}
-        let test = [0xa3, 0x01, 0x18, 0x2A, 0x02, 0x18, 0x2A, 0x01, 0x18, 0x2B];
-        let map: Result<HeaderMap, _> = serde_cbor::from_slice(&test);
-        assert!(map.is_err());
-
-        // {1: 42, 2: 42, 2: 42}
-        let test = [0xa3, 0x01, 0x18, 0x2A, 0x02, 0x18, 0x2A, 0x02, 0x18, 0x2A];
-        let map: Result<HeaderMap, _> = serde_cbor::from_slice(&test);
-        assert!(map.is_err());
-    }
+        #[test]
+        fn map_with_duplicates() {
```

Comment:
This test doesn't depend on openssl, so it should be moved outside of `mod generic`.